### PR TITLE
Smart initialization added. New class StepsGaitGenerator added. New t…

### PR DESCRIPTION
**Pull request by pajovic23**
@pajovic23
Smart initialization added. New class StepsGaitGenerator added. New terrain Steps added. Ruben's
comments added for eigen class inside the other class.
Remarks for the Steps terrain:
 - The solution can vary based on the desired speed of the robot. This could be changed by
   changing the goal, or by changing the time of the simulation.
 - There are two parameters: step_margins and swing_phase_duration, which could be also be tuned
   in order for solver to find the solution, but this cannot be changed in the runtime
 - Since the StepsGaitGenerator is used, the towr will break, if the generated are less or equal
   zero. This could happen if: - step_margins or swing_phase_durations are too high.
                               - the goal is in the middle of the terrain. (that is why the goal
                                 the goal should be always set at the end of the last step +
                                 some small margin in order for robot to finish entirely on the
                                 last step)

### CODE REVIEW

**Comment by mbjelonic** [comment]
I was just running the branch. The motion looks good. It seems like the smart initialization is not generalized with respect to the goal position. It works fine with different durations.

**Comment by pajovic23** [comment]
> I was just running the branch. The motion looks good. It seems like the smart initialization is not generalized with respect to the goal position. It works fine with different durations.

Do you mean not generalized for different goals on the y-axis?

**Comment by mbjelonic** [comment]
No with respect to the x axis.

**Comment by pajovic23** [comment]
> No with respect to the x axis.

Ok, if you mean putting the x-axis closer to the step, or before the step, then that one is not implemented, because phase duration generation is not implemented in that way, but that is something I could do and push the new commit with that one. Earlier I was only considering that we want to cross the step.

**Comment by mbjelonic** [comment]
@pajovic23 The motion in towr is looking good to me. I tried to track the motion in the MPC and it is failing right in the beginning. I can see in towr that your ground reaction forces (red arrows) are quite long (big force) and it is also not smooth, i.e., the forces are jumping around. A comparison with Vivian's version shows that this is not happening in her branch.

Your optimized trajectories:
image

**Comment by mbjelonic** [comment]
> > No with respect to the x axis.
> 
> Ok, if you mean putting the x-axis closer to the step, or before the step, then that one is not implemented, because phase duration generation is not implemented in that way, but that is something I could do and push the new commit with that one. Earlier I was only considering that we want to cross the step.

No, its fine for me. Please add this as a todo. I think it is more important to fix the vibration of the forces.

**Comment by pajovic23** [comment]
> > > No with respect to the x axis.
> > 
> > 
> > Ok, if you mean putting the x-axis closer to the step, or before the step, then that one is not implemented, because phase duration generation is not implemented in that way, but that is something I could do and push the new commit with that one. Earlier I was only considering that we want to cross the step.
> 
> No, its fine for me. Please add this as a todo. I think it is more important to fix the vibration of the forces.

Ok, I will look into that one.

**Comment by mbjelonic** [comment]
> @pajovic23 The motion in towr is looking good to me. I tried to track the motion in the MPC and it is failing right in the beginning. I can see in towr that your ground reaction forces (red arrows) are quite long (big force) and it is also not smooth, i.e., the forces are jumping around. A comparison with Vivian's version shows that this is not happening in her branch.
> 
> Your optimized trajectories:
> image

I was running your branch on flat terrain with the initialization of the normal trot and it was working fine, i.e., I can track it with the MPC and in the towr visualization the forces are reasonable and not shaky.

**Comment by pajovic23** [comment]
> > @pajovic23 The motion in towr is looking good to me. I tried to track the motion in the MPC and it is failing right in the beginning. I can see in towr that your ground reaction forces (red arrows) are quite long (big force) and it is also not smooth, i.e., the forces are jumping around. A comparison with Vivian's version shows that this is not happening in her branch.
> > Your optimized trajectories:
> > image
> 
> I was running your branch on flat terrain with the initialization of the normal trot and it was working fine, i.e., I can track it with the MPC and in the towr visualization the forces are reasonable and not shaky.

Ok, that is because my branch is not affecting the trotting gait on the flat terrain. Could you try to change the parameter force_limit_in_normal_direction_,  because right now it is taken from the vivian's version and try to put it lower, on for example 300?

**Comment by mbjelonic** [comment]
> > > @pajovic23 The motion in towr is looking good to me. I tried to track the motion in the MPC and it is failing right in the beginning. I can see in towr that your ground reaction forces (red arrows) are quite long (big force) and it is also not smooth, i.e., the forces are jumping around. A comparison with Vivian's version shows that this is not happening in her branch.
> > > Your optimized trajectories:
> > > image
> > 
> > 
> > I was running your branch on flat terrain with the initialization of the normal trot and it was working fine, i.e., I can track it with the MPC and in the towr visualization the forces are reasonable and not shaky.
> 
> Ok, that is because my branch is not affecting the trotting gait on the flat terrain. Could you try to change the parameter force_limit_in_normal_direction_, because right now it is taken from the vivian's version and try to put it lower, on for example 300?

But Vivian's version was fine when going over the step.

**Comment by pajovic23** [comment]
> > > > @pajovic23 The motion in towr is looking good to me. I tried to track the motion in the MPC and it is failing right in the beginning. I can see in towr that your ground reaction forces (red arrows) are quite long (big force) and it is also not smooth, i.e., the forces are jumping around. A comparison with Vivian's version shows that this is not happening in her branch.
> > > > Your optimized trajectories:
> > > > image
> > > 
> > > 
> > > I was running your branch on flat terrain with the initialization of the normal trot and it was working fine, i.e., I can track it with the MPC and in the towr visualization the forces are reasonable and not shaky.
> > 
> > 
> > Ok, that is because my branch is not affecting the trotting gait on the flat terrain. Could you try to change the parameter force_limit_in_normal_direction_, because right now it is taken from the vivian's version and try to put it lower, on for example 300?
> 
> But Vivian's version was fine when going over the step.

I looked into the problem with forces. Could you make sure that you are obtaining valid solution from the towr? From my tests, these types of forces could happen, but only when the towr ends up with solution which violates the contraints. These motions could look fine, and also it could happen that it only violates one constraint so you didn't notice that acutally the solution is not valid. This is happening for example when you directly pass to my terrain and optimize, the solution for that desired speed, violates the constraints and thus is giving those types of forces. 
Just change the desired time, to affect the speed, and there should always be some desired speeds for which the solution is not violating the constraints and that will produce forces without jumps or too high forces.

**Comment by mbjelonic** [comment]
> > > > > @pajovic23 The motion in towr is looking good to me. I tried to track the motion in the MPC and it is failing right in the beginning. I can see in towr that your ground reaction forces (red arrows) are quite long (big force) and it is also not smooth, i.e., the forces are jumping around. A comparison with Vivian's version shows that this is not happening in her branch.
> > > > > Your optimized trajectories:
> > > > > image
> > > > 
> > > > 
> > > > I was running your branch on flat terrain with the initialization of the normal trot and it was working fine, i.e., I can track it with the MPC and in the towr visualization the forces are reasonable and not shaky.
> > > 
> > > 
> > > Ok, that is because my branch is not affecting the trotting gait on the flat terrain. Could you try to change the parameter force_limit_in_normal_direction_, because right now it is taken from the vivian's version and try to put it lower, on for example 300?
> > 
> > 
> > But Vivian's version was fine when going over the step.
> 
> I looked into the problem with forces. Could you make sure that you are obtaining valid solution from the towr? From my tests, these types of forces could happen, but only when the towr ends up with solution which violates the contraints. These motions could look fine, and also it could happen that it only violates one constraint so you didn't notice that acutally the solution is not valid. This is happening for example when you directly pass to my terrain and optimize, the solution for that desired speed, violates the constraints and thus is giving those types of forces.
> Just change the desired time, to affect the speed, and there should always be some desired speeds for which the solution is not violating the constraints and that will produce forces without jumps or too high forces.

I tried now for a different speed and it seems to work fine. However, when I test with the default settings over the "Steps" it gives me the shaky force profile, while no constraint is violated.

I will go through the code next. It would be great if we can make the 0.205 obstacles work. :) I think for such an obstacle we have to slow down over the obstacle and explore better approaches.

File: towr/include/towr/terrain/examples/height_map_examples.h [hunk 1]
```diff
@@ -184,8 +184,37 @@ class ChimneyLR : public HeightMap {
   const double x_end2_ = x_start_+2*length_;
 };
 
+/**
+ * @brief Sample terrain with multiple steps in x-direction.
+ */
+class Steps : public HeightMap{
```
Comment (mbjelonic) [inline towr/include/towr/terrain/examples/height_map_examples.h:190]: This is only one Step and not multiple.

File: towr/src/nodes_variables.cc [hunk 2]
```diff
@@ -197,14 +201,43 @@ NodesVariables::AdvancedInititialisationEE(const VectorXd& initial_val,
 
       double goalz_terrain_ = terrain->GetHeight(goalx,goaly);
 
-      if(contact){
-        goalz = goalz_terrain_;
-        // goalz = 0.0; if use gap terrain, use the commented code line for initialization
+      //depending on the terrain, initialization is different
+      if (terrainID== HeightMap::TerrainID::StepsID){
```
Comment (mbjelonic) [inline towr/src/nodes_variables.cc:208]: Can we have the smart initialization be active for other terrains? Or, is it possible to use the gait command in towr to switch to smart initialization, i.e., with the keyboard "g" I can switch to smart initialization?

File: towr/src/nodes_variables.cc [hunk 3]
```diff
@@ -284,7 +319,53 @@ NodesVariables::AdvancedInititialisationBase(const VectorXd& initial_val,
...
+        if((goalx>(step_position-range_before_step)) && (goalx<(step_position+range_after_step))) {
+          goalz = pre_step_target_value_z + ((t_current3-start_time)*(post_step_target_value_z-pre_step_target_value_z))/time_to_cross;
+        }
+       // goalz = initial_val.z() + (t_current3*(final_val.z()-initial_val.z()))/t_total;
+        if(goalx>(step_position+range_after_step)){
+          goalz=post_step_target_value_z;
+        }
+        if(goalx<(step_position-range_before_step)){
+          goalz=pre_step_target_value_z;
+        }
```
Comment (mbjelonic) [inline towr/src/nodes_variables.cc]: Why all these commented lines? Can we remove them?

File: towr/src/quadruped_gait_generator.cc [hunk 4]
```diff
@@ -542,4 +545,68 @@ QuadrupedGaitGenerator::GetStrideLimp () const
...
+    durations.emplace_back(touchdown_time.at(i)-swing_phase_duration-last_step_touchdown_time);
+    durations.emplace_back(swing_phase_duration);
+    last_step_touchdown_time=touchdown_time.at(i);
+  }
+  durations.emplace_back(total_t_integer-last_step_touchdown_time);
+  for(auto t=durations.begin(); t!=durations.end(); ++t){
+    *t=*t/100;
+    //std::cout << "ee " << ee << " phase duration:" <<  *t << std::endl;
+  }
+  return durations;
```
Comment (mbjelonic) [inline towr/src/quadruped_gait_generator.cc:607]: Why all these commented lines? Can we remove it?

File: towr/src/steps_gait_generator.cc [hunk 5]
```diff
@@ -0,0 +1,75 @@
+#include <towr/initialization/steps_gait_generator.h>
+#include <iostream>
+
+namespace towr{
+
+StepsGaitGenerator::StepsGaitGenerator(double des_v_x, std::shared_ptr<towr::HeightMap> terrain){
```
Comment (mbjelonic) [inline towr/src/steps_gait_generator.cc:6]: Can we have this be general with any terrain? I.e. we can toggle the smart initialization?

**Review by mbjelonic — APPROVED** [review]

**Pull request merged**